File: backend/src/main/java/com/inbyuso/auth/service/BannerService.java
package com.inbyuso.auth.service;

import com.inbyuso.auth.dto.BannerRequest;
import com.inbyuso.auth.dto.BannerResponse;
import com.inbyuso.auth.entity.Banner;
import com.inbyuso.auth.repository.BannerRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class BannerService {

    private final BannerRepository bannerRepository;

    @Transactional(readOnly = true)
    public List<BannerResponse> getActiveBanners() {
        return bannerRepository.findByActiveTrueOrderBySortOrderAsc()
                .stream().map(BannerResponse::from).toList();
    }

    @Transactional(readOnly = true)
    public List<BannerResponse> getAllBanners() {
        return bannerRepository.findAllByOrderBySortOrderAsc()
                .stream().map(BannerResponse::from).toList();
    }

    @Transactional
    public BannerResponse create(BannerRequest request) {
        Banner banner = Banner.builder()
                .title(request.title())
                .imageUrl(request.imageUrl())
                .linkUrl(request.linkUrl())
                .sortOrder(request.sortOrder())
                .build();
        return BannerResponse.from(bannerRepository.save(banner));
    }

    @Transactional
    public BannerResponse update(UUID id, BannerRequest request) {
        Banner banner = findById(id);
        banner.update(request.title(), request.imageUrl(), request.linkUrl(), request.sortOrder());
        return BannerResponse.from(banner);
    }

    @Transactional
    public BannerResponse toggleActive(UUID id) {
        Banner banner = findById(id);
        banner.toggleActive();
        return BannerResponse.from(banner);
    }

    @Transactional
    public void delete(UUID id) {
        bannerRepository.delete(findById(id));
    }

    private Banner findById(UUID id) {
        return bannerRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "배너를 찾을 수 없습니다"));
    }
}

File: frontend/src/lib/bannerService.ts
import { apiClient } from './api'
import { ApiResponse } from '@/types/auth'

export interface Banner {
  id: string
  title: string
  imageUrl: string
  linkUrl?: string
  sortOrder: number
  active: boolean
}

export interface BannerRequest {
  title: string
  imageUrl: string
  linkUrl?: string
  sortOrder: number
}

export const bannerService = {
  async getActiveBanners(): Promise<Banner[]> {
    const res = await apiClient.get<ApiResponse<Banner[]>>('/banners')
    return res.data.data ?? []
  },

  async getAllBanners(): Promise<Banner[]> {
    const res = await apiClient.get<ApiResponse<Banner[]>>('/banners/admin')
    return res.data.data ?? []
  },

  async create(data: BannerRequest): Promise<Banner> {
    const res = await apiClient.post<ApiResponse<Banner>>('/banners/admin', data)
    return res.data.data!
  },

  async update(id: string, data: BannerRequest): Promise<Banner> {
    const res = await apiClient.put<ApiResponse<Banner>>(`/banners/admin/${id}`, data)
    return res.data.data!
  },

  async toggle(id: string): Promise<Banner> {
    const res = await apiClient.patch<ApiResponse<Banner>>(`/banners/admin/${id}/toggle`)
    return res.data.data!
  },

  async delete(id: string): Promise<void> {
    await apiClient.delete(`/banners/admin/${id}`)
  },
}
